File: libs/openapi-codegen/src/ir/names.rs
//! Identifier casing and collision-safe naming for generated TypeScript.

use std::collections::BTreeSet;

/// Split an arbitrary spec key into word tokens on non-alphanumeric
/// boundaries and camelCase humps.
fn words(input: &str) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    let mut cur = String::new();
    let mut prev_lower = false;
    for ch in input.chars() {
        if ch.is_alphanumeric() {
            if ch.is_uppercase() && prev_lower && !cur.is_empty() {
                out.push(std::mem::take(&mut cur));
            }
            cur.push(ch);
            prev_lower = ch.is_lowercase() || ch.is_numeric();
        } else if !cur.is_empty() {
            out.push(std::mem::take(&mut cur));
            prev_lower = false;
        }
    }
    if !cur.is_empty() {
        out.push(cur);
    }
    out
}

fn capitalize(word: &str) -> String {
    let mut chars = word.chars();
    match chars.next() {
        Some(first) => first.to_uppercase().collect::<String>() + &chars.as_str().to_lowercase(),
        None => String::new(),
    }
}

/// PascalCase a spec key (`pet_category` -> `PetCategory`). Falls back to a
/// leading underscore when the result would not start with a letter.
pub fn to_pascal(input: &str) -> String {
    let mut s: String = words(input).iter().map(|w| capitalize(w)).collect();
    if s.is_empty() {
        s = "Anonymous".to_string();
    }
    if !s.chars().next().map(|c| c.is_alphabetic()).unwrap_or(false) {
        s.insert(0, '_');
    }
    s
}

/// camelCase a spec key (`PetCategory` -> `petCategory`).
pub fn to_camel(input: &str) -> String {
    let pascal = to_pascal(input);
    let mut chars = pascal.chars();
    match chars.next() {
        Some(first) => first.to_lowercase().collect::<String>() + chars.as_str(),
        None => pascal,
    }
}

/// snake_case a spec key (`PetCategory` -> `pet_category`). Used for Python and
/// Rust function/parameter identifiers. Falls back to a leading underscore when
/// the result would not start with a letter or underscore.
pub fn to_snake(input: &str) -> String {
    let s = words(input)
        .iter()
        .map(|w| w.to_lowercase())
        .collect::<Vec<_>>()
        .join("_");
    if s.is_empty() {
        "field".to_string()
    } else if !s
        .chars()
        .next()
        .map(|c| c.is_alphabetic() || c == '_')
        .unwrap_or(false)
    {
        format!("_{s}")
    } else {
        s
    }
}

/// True when `s` is a valid bare TypeScript object-property / identifier.
pub fn is_ident(s: &str) -> bool {
    let mut chars = s.chars();
    match chars.next() {
        Some(c) if c.is_ascii_alphabetic() || c == '_' || c == '$' => {}
        _ => return false,
    }
    chars.all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '$')
}

/// Object-literal property key: bare when a valid identifier, else quoted.
pub fn prop_key(key: &str) -> String {
    if is_ident(key) {
        key.to_string()
    } else {
        format!("\"{}\"", key.replace('\\', "\\\\").replace('"', "\\\""))
    }
}

/// Member access against a `params` object: dotted for identifiers, bracketed
/// (string-keyed) otherwise.
pub fn param_access(name: &str) -> String {
    if is_ident(name) {
        format!("params.{name}")
    } else {
        format!(
            "params[\"{}\"]",
            name.replace('\\', "\\\\").replace('"', "\\\"")
        )
    }
}

/// Allocates unique names, suffixing `_2`, `_3`, ... on collision.
#[derive(Debug, Default)]
pub struct NameRegistry {
    used: BTreeSet<String>,
}

impl NameRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn unique(&mut self, base: &str) -> String {
        let base = if base.is_empty() { "anonymous" } else { base };
        if self.used.insert(base.to_string()) {
            return base.to_string();
        }
        let mut n = 2;
        loop {
            let candidate = format!("{base}_{n}");
            if self.used.insert(candidate.clone()) {
                return candidate;
            }
            n += 1;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pascal_and_camel() {
        assert_eq!(to_pascal("pet_category"), "PetCategory");
        assert_eq!(to_pascal("pet-category"), "PetCategory");
        assert_eq!(to_camel("PetCategory"), "petCategory");
        assert_eq!(to_camel("list_pets"), "listPets");
    }

    #[test]
    fn pascal_handles_leading_non_alpha() {
        assert_eq!(to_pascal("123abc"), "_123abc");
    }

    #[test]
    fn registry_suffixes_on_collision() {
        let mut reg = NameRegistry::new();
        assert_eq!(reg.unique("Pet"), "Pet");
        assert_eq!(reg.unique("Pet"), "Pet_2");
        assert_eq!(reg.unique("Pet"), "Pet_3");
    }

    #[test]
    fn ident_detection() {
        assert!(is_ident("petId"));
        assert!(!is_ident("X-Request-Id"));
        assert_eq!(prop_key("X-Request-Id"), "\"X-Request-Id\"");
        assert_eq!(prop_key("petId"), "petId");
    }
}

File: libs/openapi-codegen/tech-design/src/openapi_codegen/domain/names.py
from __future__ import annotations


def words(text: str) -> tuple[str, ...]:
    emitted: list[str] = []
    current: list[str] = []
    prev_lower = False

    for ch in text:
        if ch.isalnum():
            if ch.isupper() and prev_lower and current:
                emitted.append("".join(current))
                current = []
            current.append(ch)
            prev_lower = ch.islower() or ch.isnumeric()
        else:
            if current:
                emitted.append("".join(current))
                current = []
            prev_lower = False

    if current:
        emitted.append("".join(current))

    return tuple(emitted)


def capitalize(word: str) -> str:
    if not word:
        return ""
    return word[0].upper() + word[1:].lower()


def to_pascal(text: str) -> str:
    s = "".join(capitalize(w) for w in words(text))
    if s == "":
        s = "Anonymous"
    if not s[0].isalpha():
        s = "_" + s
    return s


def to_camel(text: str) -> str:
    p = to_pascal(text)
    return p[0].lower() + p[1:]


def to_snake(text: str) -> str:
    s = "_".join(w.lower() for w in words(text))
    if s == "":
        return "field"
    if not s[0].isalpha() and s[0] != "_":
        return "_" + s
    return s


def is_ident(text: str) -> bool:
    if not text:
        return False
    c0 = text[0]
    if not (("a" <= c0 <= "z") or ("A" <= c0 <= "Z") or c0 == "_" or c0 == "$"):
        return False
    for ch in text[1:]:
        if not (
            ("a" <= ch <= "z")
            or ("A" <= ch <= "Z")
            or ("0" <= ch <= "9")
            or ch == "_"
            or ch == "$"
        ):
            return False
    return True


def escape_string_literal(text: str) -> str:
    return text.replace("\\", "\\\\").replace('"', '\\"')


def prop_key(key: str) -> str:
    if is_ident(key):
        return key
    return '"' + escape_string_literal(key) + '"'


def param_access(name: str) -> str:
    if is_ident(name):
        return "params." + name
    return 'params["' + escape_string_literal(name) + '"]'


class NameRegistry:
    def __init__(self) -> None:
        self._used: set[str] = set()

    def unique(self, base: str) -> str:
        if base == "":
            base = "anonymous"
        if base not in self._used:
            self._used.add(base)
            return base
        n = 2
        while True:
            candidate = f"{base}_{n}"
            if candidate not in self._used:
                self._used.add(candidate)
                return candidate
            n += 1

    def taken(self) -> frozenset[str]:
        return frozenset(self._used)
